static/js/lib/ckeditor/turndown.ts

```diff
@@ -111,3 +111,121 @@ itemRule.replacement = (
+ *
+ * Issue #1
+ * -----------
+ * The default behavior can result in HTML tags beginning a markdown line that
+ * would otherwise be a Markdown paragraph. For example:
+ * ```js
+ * const tds = new TurndownService()
+ * tds.keep(["sup"])
+ * tds.turndown("<p><sup>2</sup> Hello <strong>world</strong>!</p>")
+ * // => "<sup>2</sup> Hello **world**!"
+ * ```
+ * The output above is bad because when an HTML tag begins a markdown line, it
+ * begins a *block* of markdown, and subsequent characters on that line are
+ * NOT treated as Markdown. Hence the `**` above would be rendered as literal
+ * asterisks, not as bold.
+ *
+ * Issue #2
+ * ------------
+ * The default behavior is that the *content* of an HTML tag is not converted
+ * to Markdown. This is problematic for tags that contain shortcode HTML. For
+ * example, `<sup><a class="resource-link" data-uuid ... >2</a></sup>` should
+ * really needs its content to be converted to markdown in order for Hugo to
+ * recognize it as a shortcode.
+ */
+class TurndownHtmlHelpers {
+  private turndownInstance: Turndown
+
+  constructor(turndown: Turndown) {
+    this.turndownInstance = turndown
+  }
+
+  /**
+   * This function improves upon Turndown's default rule for including raw HTML
+   * tags in Markdown. In particular, it:
+   *  - ensures that the child content of an inline HTML node is also converted, which
+   *    we need in case the child content contains shortcode.
+   *  - ensures that an inline HTML tag never begins a line of Markdown.
+   *
+   * Should be used alongside {@link TurndownHtmlHelpers.turndown}
+   */
+  keepReplacer: Turndown.ReplacementFunction = (content, node) => {
+    /**
+     * "isBlock" is a Turndown-specific addition to the DOM node interface.
+     * It appears to be part of their public API in that the author recommends
+     * using it in several issues.
+     */
+    const el = node as HTMLElement & { isBlock: boolean }
+
+    if (el.isBlock) {
+      throw new Error("Inclusion of block content not yet supported.")
+    }
+
+    /**
+     * Now convert the node's child content to Markdown.
+     * This addresses Issue #2 above.
+     */
+    const clone = el.cloneNode() as HTMLElement
+    clone.innerHTML = this.turndownInstance.turndown(el.innerHTML)
+
+    /**
+     * To address Issue #1, we'll mark places where HTML is inserted. Later, we
+     * can remove all the markings except the ones that begin a new line.
+     */
+    return `<raw_inline></raw_inline>${clone.outerHTML}`
+  }
+
+  /**
+   * Convert HTML to markdown and insert a zero-width space before any HTML
+   * tag that begins what would otherwise be a Markdown paragraph. Should
+   * be used alongside {@link TurndownHtmlHelpers.keepReplacer}.
+   *
+   * See {@link TurndownHtmlHelpers} for more details.
+   */
+  turndown = (html: string) =>
+    this.turndownInstance
+      .turndown(html)
+      .replace(/^[ ]*<raw_inline><\/raw_inline>/g, "\u200b")
```

Comment:
We've used techniques similar to this with our Turndown code in `ocw-to-hugo`, before we switched to shortcode tables. We would leave markers to later replace with pipes to handle constructing tables in markdown. This is a bit different of an issue, but the concept remains the same: leave a specific marker that can be replaced later in the process.

As far as the zero width space, I think it's probably fine as long as it doesn't introduce any practical formatting / word wrapping issues with CSS on the built site. Since it's being inserted at the beginning of the line, I don't see it being a problem. I was trying to reproduce this by putting a subscript at the beginning of a line and saving it, but I didn't even see the character in the raw markdown pushed up to Github. Either way, I don't think Hugo will render it right?

Reply:
This was an error on my part and fixed in https://github.com/mitodl/ocw-studio/pull/1588/commits/1539f1942931827eff2f590466595a791efba831. The zwsp was only being inserted at the beginning of the text, not at the beginning of a line.

In the example you shared on slack (https://github.com/gumaerc-testorg/5.13-fall-2003/blob/main/content/pages/syllabus.md) ... If you had done `x_2 blah blah **foo**` I expect that `**foo**` would have shown up with literal asterisks, not bold. Now it should show bold.